Fix login as other user

mkangia opened the pull request:
https://manage.dimagi.com/default.asp?278446#1508216

There is a js error when you visit https://www.commcarehq.org/hq/admin/auth_as/ which happns [here](https://github.com/dimagi/commcare-hq/blob/9d031cc561c5c3a956654a16891ae08da75072c7/corehq/apps/reports/static/reports/js/tabular.js#L63) because there is no `data`. This is happening for XHR requests to urls like "https://i.kissmetrics.com/i.js?_=1530283757604"

tabular.js?version=9233b57:63 Uncaught TypeError: Cannot read property 'slug' of undefined
    at HTMLDocument.<anonymous> (tabular.js?version=9233b57:63)
    at HTMLDocument.dispatch (jquery.min.js?version=1055018:3)
    at HTMLDocument.q.handle (jquery.min.js?version=1055018:3)
    at Object.trigger (jquery.min.js?version=1055018:4)
    at A (jquery.min.js?version=1055018:4)
    at HTMLScriptElement.c (jquery.min.js?version=1055018:4)
    at HTMLScriptElement.dispatch (jquery.min.js?version=1055018:3)
    at HTMLScriptElement.q.handle (jquery.min.js?version=1055018:3)

I don't think that js block needs to run for this page. Looks like js_options is a requirement of using the base view `hqadmin/hqadmin_base_report.html` so just added that as empty and now it returns out of that block.

I am not sure if there is a better way to address this like change the template view or update that JS to run `renderPage(data.slug, data.report_table_js_options);` only when there is data but that might bring unwanted changes or ignore real errors.

mkangia committed:
Add js_options to satisfy requirement for admin base report view

orangejenny commented:
The error is coming out of tabular.js, which is included because this page inherits from hqadmin_base_report...but this page isn't a report and doesn't use tabular data. Why not inherit from two_column.html or base_section.html instead?

It's also not clear to me that this fixes the issue. When I check out your branch, I still see the behavior Kai reported (nothing happens if you fill in email, 404 if you fill in both email and domain).

mkangia committed:
add a parent template for user admin; then use that instead

mkangia committed:
revert changes done earlier to compensate for js options

mkangia committed:
show errors to user

mkangia committed:
dont manipulate the url

there should be no need to manipulate the url since cripsy forms would send the username and domain in form data

mkangia committed:
check for commcare user only when it should be one

mkangia committed:
dont need the js anymrore

mkangia committed:
remove test for allowing just mobile user

mkangia commented:
Thank you @orangejenny 

Did few things:
1. Added `UserAdministration` view and corresponding template to extract things needed for its children views from `hqadmin_base_report`. Removed any reference to `report` in this template as well.
2. Made some improvements: Render error messages and removed changes to url which eventually resulted in removing the `authenticate_as` js since it didn't do anything else.

The one change i made which changes the functionality is allowing login as web user as well. It was not clear if it was intended to be kept out to. Don't think it would do a harm.

orangejenny commented on corehq/apps/hqadmin/views/users.py:
```diff
@@ -93,6 +93,11 @@ def post(self, request, *args, **kwargs):
             request.user.backend = 'django.contrib.auth.backends.ModelBackend'
             login(request, request.user)
             return HttpResponseRedirect('/')
+        all_errors = form.errors.pop('__all__', None)
```
This is a nice bit of improvement.

orangejenny commented on corehq/apps/hqadmin/static/hqadmin/js/authenticate_as.js:
```diff
@@ -1,12 +0,0 @@
-/* globals hqDefine */
```
Sweet, always happy to see code deleted.

esoergel commented on corehq/apps/hqadmin/forms.py:
```diff
@@ -51,8 +51,9 @@ def clean(self):
-            raise forms.ValidationError("User '{}' is not a CommCareUser".format(username))
+        if domain or '.commcarehq.org' in username:
+            if not user.is_commcare_user():
+                raise forms.ValidationError("User '{}' is not a CommCareUser".format(username))
```
I think this should use `CouchUser.get_by_username` if its expected to sometimes return `WebUser`s.

mkangia replied:
reverted to allow login as a mobile user only, so not taking any action for this.

esoergel commented on corehq/apps/hqadmin/tests/test_authenticate_as.py:
```diff
@@ -55,14 +55,6 @@ def test_invalid_data(self):
-        }
-        form = AuthenticateAsForm(data)
-        self.assertFalse(form.is_valid())
-        self.assertTrue('not a CommCareUser' in str(form.errors))
```
You could turn this into a test validating that logging in as a web user works

mkangia replied:
reverted.

esoergel commented on corehq/apps/hqadmin/views/users.py:
```diff
@@ -89,10 +93,15 @@ def post(self, request, *args, **kwargs):
+        if all_errors:
+            messages.error(request, ','.join(all_errors))
+        if form.errors:
+            messages.error(request, form.errors)
```
Could you instead render the form errors in the html?  Django has good support for this kind of thing:
https://docs.djangoproject.com/en/1.11/topics/forms/#rendering-form-error-messages

I expect crispy_forms does too - doesn't it automatically render both field specific errors and `non_field_errors`?

mkangia replied:
Interesting... i tried rendering errors with fields but that didn't work. Going to not worry about it for this and leave it the way it is. Sorry.

mkangia committed:
Revert "remove test for allowing just mobile user"

This reverts commit 63f99d3c12ed77cb704cd7f9b90970eb6004d462.

mkangia committed:
make it clear that its only for mobile worker

mkangia committed:
clean up

- dont edit cleaned_data
- extract full_username and use that instead
- dont need different error messages, keep it simple

mkangia committed:
fix all views including audited_views

mkangia committed:
bring back js and update url to have proper audit

the way audit is set up currently, to be able to track with user it was tried to login as, username and domain needs to be in the url. So put that but use root_page_url to avoid issues during append

mkangia committed:
just add a link to the doc to view more details if needed

esoergel commented on corehq/ex-submodules/auditcare/views.py:
```diff
@@ -114,31 +114,31 @@ def model_instance_history(request, model_name, model_uuid, *args, **kwargs):
-    return render_to_response('auditcare/model_changes.html', context)
+    model_dict = dict((x['value'][0], x['value']) for x in vals)
+    context = {'model_dict': model_dict}
+    return render(request, 'auditcare/model_changes.html', context)
```
wow, some old-school Django in this file

mkangia replied:
tell me about it.. :)

mkangia commented:
thanks for finding that PR @esoergel .. 

That answered few things for me... so i have updated this. commit messages should be clear but just to sum up:
1. allow login only as a mobile worker and make it more clear
2. bring back the js file (sorry @orangejenny) for audit to work correctly. Just fixed that to pick the base/root url instead of current url which was causing the error as mentioned in the ticket.
3. this view and some others were actually just broken. Looks like we don't use this at all. So fixed those.
4. Also made some random improvements
5. going to add some comments around so that someone naive like me does not mess it up.

orangejenny commented:
> bring back the js file

No worries, js has its place ;)

> going to add some comments around so that someone naive like me does not mess it up.

A+

mkangia committed:
add some documentation/explanation
